src/JsonSchema/SchemaFactory.php

```diff
@@ -194,8 +201,18 @@ private function buildPropertySchema(Schema $schema, string $definitionName, str
             }
 
             $subSchema = $this->buildSchema($className, $format, Schema::TYPE_OUTPUT, null, $subSchema, $serializerContext + [self::FORCE_SUBSCHEMA => true], false);
-            $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
-            // prevent "type" and "anyOf" conflict
+            if ($isCollection) {
+                $propertySchema['items']['$ref'] = $subSchema['$ref'];
+                unset($propertySchema['items']['type']);
+                break;
+            }
+
+            if ($type->isNullable()) {
+                $propertySchema['anyOf'] = [['$ref' => $subSchema['$ref']], ['type' => 'null']];
+            } else {
+                $propertySchema['oneOf'] = [['$ref' => $subSchema['$ref']]];
```

Comment:
Useless `oneOf` in this case, `$ref` is enough (cf. https://json-schema.org/understanding-json-schema/structuring.html#ref)